modules/text2vec-upstage/clients/upstage.go

```diff
@@ -0,0 +1,252 @@
+	cfg moduletools.ClassConfig,
+) (*modulecomponents.VectorizationResult, error) {
+	config := v.getVectorizationConfig(cfg)
+	return v.vectorize(ctx, input, config.Model, config)
+}
+
+func (v *vectorizer) getVectorizationConfig(cfg moduletools.ClassConfig) ent.VectorizationConfig {
+	icheck := ent.NewClassSettings(cfg)
+	return ent.VectorizationConfig{
+		Model:   icheck.Model(),
+		BaseURL: icheck.BaseURL(),
+	}
+}
+
+func (v *vectorizer) vectorize(ctx context.Context, input []string, model string, config ent.VectorizationConfig) (*modulecomponents.VectorizationResult, error) {
+	body, err := json.Marshal(v.getEmbeddingsRequest(input, model))
+	if err != nil {
+		return nil, errors.Wrap(err, "marshal body")
+	}
+
+	endpoint, err := v.buildUrlFn(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "join upstage API host and path")
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", endpoint,
+		bytes.NewReader(body))
+	if err != nil {
+		return nil, errors.Wrap(err, "create POST request")
+	}
+	apiKey, err := v.getApiKey(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "API Key")
+	}
+	req.Header.Add(v.getApiKeyHeaderAndValue(apiKey))
+	req.Header.Add("Content-Type", "application/json")
+
+	res, err := v.httpClient.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "send POST request")
+	}
+	defer res.Body.Close()
+
+	bodyBytes, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "read response body")
+	}
+
+	var resBody embedding
+	if err := json.Unmarshal(bodyBytes, &resBody); err != nil {
+		return nil, errors.Wrap(err, "unmarshal response body")
+	}
+
+	if res.StatusCode != 200 || resBody.Error != nil {
+		return nil, v.getError(res.StatusCode, resBody.Error)
+	}
+
+	texts := make([]string, len(resBody.Data))
+	embeddings := make([][]float32, len(resBody.Data))
+	for i := range resBody.Data {
+		index := resBody.Data[i].Index
+		texts[index] = resBody.Data[i].Object
+		embeddings[index] = resBody.Data[i].Embedding
+	}
+
+	return &modulecomponents.VectorizationResult{
+		Text:       texts,
+		Dimensions: len(resBody.Data[0].Embedding),
+		Vector:     embeddings,
+	}, nil
+}
+
+func (v *vectorizer) getError(statusCode int, resBodyError *upstageApiError) error {
+	endpoint := "upstage API"
+	if resBodyError != nil {
+		return fmt.Errorf("connection to: %s failed with status: %d error: %v", endpoint, statusCode, resBodyError.Message)
+	}
+	return fmt.Errorf("connection to: %s failed with status: %d", endpoint, statusCode)
+}
+
+func (v *vectorizer) getEmbeddingsRequest(input []string, model string) embeddingsRequest {
+	return embeddingsRequest{Input: input, Model: model}
+}
+
+func (v *vectorizer) getApiKeyHeaderAndValue(apiKey string) (string, string) {
+	return "Authorization", fmt.Sprintf("Bearer %s", apiKey)
+}
+
+func (v *vectorizer) getApiKey(ctx context.Context) (string, error) {
+	var apiKey, envVar string
+
+	apiKey = "X-upstage-Api-Key"
```

Comment:
here should be `X-Upstage-Api-Key`

Reply:
you should also use `modulecomponents.GetValueFromContext` method here first instead of `getApiKeyFromContext` (which you can remove in favor of `GetValueFromContext` method).

The idea is that a user should always be able to provide the APIKEY using headers no matter if `upstageApiKey` has this value defined. We want always be able to override it